File: ml-service/app/services/gemini_client.py
"""Gemini (Google) integration for ML service."""

from __future__ import annotations

import httpx
from typing import Optional


class GeminiClient:
    def __init__(self, api_key: str, model: str | None = None, timeout_seconds: int = 120) -> None:
        self.api_key = api_key
        self.model = model or "gemini-2.5-flash"
        self.timeout_seconds = timeout_seconds

    async def generate(self, prompt: str, response_mime: str = "text/plain") -> str:
        endpoint = (
            f"https://generativelanguage.googleapis.com/v1beta/models/{self.model}:generateContent"
            f"?key={httpx.utils.quote(self.api_key)}"
        )
        payload = {
            "contents": [{"role": "user", "parts": [{"text": prompt}]}],
            "generationConfig": {"temperature": 0.4, "responseMimeType": response_mime},
        }
        timeout = httpx.Timeout(10.0, read=self.timeout_seconds)
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.post(endpoint, json=payload)
            resp.raise_for_status()
            j = resp.json()
            # Extract text from candidates -> content -> parts
            parts = (
                j.get("candidates", []) and j["candidates"][0].get("content", {}).get("parts", [])
            )
            if not parts:
                # fallback: try to return raw as string
                return str(j)
            text = "\n".join([p.get("text", "") for p in parts]).strip()
            return text

File: backend/src/integrations/gemini/geminiClient.js
import env from '../../config/env.js';

function extractText(responseJson = {}) {
  const parts = responseJson?.candidates?.[0]?.content?.parts || [];
  return parts.map((part) => part?.text || '').join('\n').trim();
}

async function requestGemini(prompt, { responseMimeType = 'text/plain' } = {}) {
  if (!env.geminiApiKey) {
    return { text: prompt, source: 'fallback' };
  }

  const model = env.geminiModel || 'gemini-2.5-flash';
  const endpoint = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${encodeURIComponent(env.geminiApiKey)}`;
  const response = await fetch(endpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
      generationConfig: {
        temperature: 0.4,
        responseMimeType,
      },
    }),
  });

  if (!response.ok) {
    const details = await response.text().catch(() => '');
    throw new Error(`Gemini request failed (${response.status}): ${details || response.statusText}`);
  }

  const json = await response.json();
  return {
    text: extractText(json),
    raw: json,
    source: 'gemini',
  };
}

export async function geminiRequest(prompt) {
  return requestGemini(prompt);
}

export async function geminiJsonRequest(prompt) {
  return requestGemini(prompt, { responseMimeType: 'application/json' });
}
